Add support for checking firmware version of network interface cards.

jovial opened the pull request:
A new HardwareManager was added which provides a verify_nic_firmware
cleaning step. The list of cards to check is set through the nic_firmware
property in the extra field of the Ironic node info.

jovial committed:
Add support for checking firmware version of network interface cards.

A new HardwareManager was added which provides a verify_nic_firmware
cleaning step. The list of cards to check is set through the nic_firmware
property in the extra field of the Ironic node info.

jovial commented:
Looking for some feedback to determine if the general approach is OK.

jovial committed:
removed some unused variables

jovial committed:
Added SystemNICHardwareManager to list of entry points

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+from oslo_concurrency import processutils
+from oslo_log import log
+
+UEVENT_PCI_SLOT_NAME_PREFIX = "PCI_SLOT_NAME="
```
nit: could prefix this with _ to indicate it's private

jovial replied:
fair point

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+    if os.path.isabs(path):
+        raise ValueError("path must be relative")
+    dirname = os.path.dirname(path)
+    if dirname == "":
```
Maybe I missed something but could you just do something like return path.split('/')[0]?

jovial replied:
I was trying to keep path handling generic, but maybe using `split` with `os.sep` would be ok.

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+
+        :returns: HardwareSupport level for this manager.
+        """
+        # This should work for anything which supports ethtool, lspci
```
We can make sure these packages are installed in the DIB element:

https://github.com/stackhpc/stackhpc-image-elements/blob/master/elements/stackhpc-ipa-hardware-managers/package-installs.yaml

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+    def _get_ethtool_output(*args, **kwargs):
+        return _get_ethtool_output(*args, **kwargs)
+
+    def _get_nic_firmware_versions(self, vendor_id, device_id):
```
This function is also factored out above the class.

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+            LOG.warning(
+                'NIC firmware property has not been set. No firmware will be '
+                'verified')
+            return True
```
Seems sensible. I guess the alternative could be to fail cleaning if no firmware is set, unless a flag ```skip_nic_firmware_check``` is set. For environments where you want to check that all firmware is as expected that might help to enforce good practice. Anyone else have an opinion?

jovial replied:
I thought the absence of that data structure would indicate that you didn't want to check the firmware, but we could add a flag if you think that would be better.

dougszumski replied:
Yeah - I guess it's a question of do we want to force people to opt out by explicitly having to set a flag, or let them opt in. @oneswig / @markgoddard ?

markgoddard replied:
what did we do with system_bios?

jovial replied:
We have a separate flag for disabling the check - I'll change so this is consistent.

dougszumski commented on stackhpc_ipa_hardware_managers/tests/unit/test_system_nic.py:
```diff
@@ -0,0 +1,149 @@
+# Copyright (c) 2017 StackHPC Ltd.
```
nit: s/2017/2018

dougszumski commented on stackhpc_ipa_hardware_managers/tests/unit/test_system_nic.py:
```diff
@@ -0,0 +1,149 @@
+    return {
+        'extra': {
+            'nic_firmware': [
+                {
```
Is it worth adding a test for multiple NIC firmwares?

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+                 'abortable': True}]
+
+    def _get_lspci_output(*args, **kwargs):
+        return _get_lspci_output(*args, **kwargs)
```
It might be more canonical to leave these here as static methods. I don't think the testing should look too different. Example:

https://github.com/stackhpc/stackhpc-monasca-agent-plugins/blob/master/stackhpc_monasca_agent_plugins/tests/unit/checks/test_slurm.py

jovial replied:
Yeah, you are right, that is mess, but I don't like the idea of mocking those methods as the test becomes dependent on the implementation. What I propose is to test the parsing separately and make the SystemNICHardwareManager tests completely independent of the parsing logic (assume it works). Thoughts?

dougszumski replied:
Yeah, fair point about not making the tests depend on the implementation / testing private methods.

markgoddard replied:
nit: I'm fine with using global functions, but why not call them directly? Looks like its to enable testing using a mock. It's pretty common practice in python (and OpenStack in particular) to use the mock library to do that sort of thing, and it allows you to check error paths. The test is dependent upon the implementation anyway due to the mock object.

dougszumski reviewed (commented):
Good stuff!

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+
+
+def _get_lspci_output(vendor_id=None, device_id=None, **kwargs):
+    device_filter_arg = ''
```
It's not worth changing, but in general it's better to pick either " or ' consistently unless using the other makes sense for quoting.

`"'"`
`'"'`

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+    if vendor_id or device_id:
+        device_filter_arg = "-d " + _get_device_filter_string(vendor_id,
+                                                              device_id)
+    lspci_output = ''
```
nit: No need to initialise as we can't get to return without setting it.

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+        lspci_output, _e = utils.execute(
+            "lspci -vmmnD {filter}".format(filter=device_filter_arg), **kwargs)
+    except (processutils.ProcessExecutionError, OSError):
+        raise errors.CleaningError("Unable to run lspci")
```
This error might not be that helpful to a user. How about adding a little context such as 'when checking NIC firmware'. It may also be possible to get the command stderr from the exception object.

jovial replied:
The message now looks like: 
```
ironic_python_agent.errors.CleaningError: Clean step failed: The command: lspci failed to execute when performing NIC firmware check. Unexpected error while running command.
Command: sudo ironic-rootwrap /etc/ironic/rootwrap.conf lspci -vmmnD 
Exit code: 1
Stdout: u''
Stderr: u'sudo: no tty present and no askpass program specified\n'
```
at least in the case of `processutils.ProcessExecutionError`

jovial committed:
update copyright

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+    return lspci_output
+
+
+def _parse_lspci_output(lspci_output):
```
nit: In text parsing code it's helpful to have a little description of the expected format, or a pointer to some docs.

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+            return value
+
+
+def _get_expected_property(node, node_property):
```
This isn't the node we're checking AFAICT.

jovial replied:
I have renamed this `firmware_matcher` (regex style) for want of a better name

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+    return expected_property
+
+
+def _get_nic_firmware_versions(vendor_id, device_id):
```
Could use a note describing the returned data format. Looks like it's a dict mapping interface name to firmware version.

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,241 @@
+                'verified')
+            return True
+
+        firmwares = node["extra"]["nic_firmware"]
```
The data format for this field needs to be documented somewhere.

markgoddard commented on stackhpc_ipa_hardware_managers/tests/unit/test_system_nic.py:
```diff
@@ -0,0 +1,149 @@
+        return f.read()
+
+
+class TestSystemNIC(unittest.TestCase):
```
Needs more testing of failure code paths.

markgoddard reviewed (changes requested):
Thanks Will

jovial committed:
prefix private variable

jovial committed:
use global functions directly

jovial committed:
be more consistent with quotes

jovial committed:
simplify path manipulation

jovial committed:
improve error messages

jovial committed:
use a consistent name for firmware matching criteria

jovial committed:
improve docs

jovial committed:
document process_firmware_matcher

jovial committed:
document _parse_lspci_output

jovial committed:
more documentation improvements

jovial committed:
improve docs for verify_nic_firmware

jovial committed:
Firmware check now has to be explicitly disabled

This is to ensure consistency with system_bios. It is now an error if
the nic_firmware property is missing and the disable_nic_firmware_check
flag is not set.

jovial committed:
test missing fields on matching criteria

jovial committed:
improve testing of firmware matching criteria structure

jovial committed:
conform to pep8

jovial committed:
improve handling of multiple mismatching devices

jovial committed:
fix unaligned indent

jovial committed:
Removed dependency on lspci (pciutils)

jovial committed:
pep8 fix

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Test coverage is a lot better:
```
$ tox -e cover
<snip>

cover runtests: commands[3] | coverage report --omit=*tests*
Name                                            Stmts   Miss  Cover
-------------------------------------------------------------------
stackhpc_ipa_hardware_managers/__init__.py          0      0   100%
stackhpc_ipa_hardware_managers/system_bios.py      60      4    93%
stackhpc_ipa_hardware_managers/system_nic.py      145     24    83%
-------------------------------------------------------------------
TOTAL                                             205     28    86%
cover runtests: commands[4] | coverage html -d ./cover --omit=*tests*
______________________________________________________ summary _______________________________________________________
  cover: commands succeeded
  congratulations :)
```

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+
+        :returns: HardwareSupport level for this manager.
+        """
+        # This should work for anything which supports ethtool
```
Did you update the DIB element now that we don't need lspci?

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+
+        return successes
+
+    def process_expected_versions(self, pci_matcher_lookup_table):
```
nit: Should this be private?

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+        the values of all fields should be strings. An example is given below::
+
+            [
+                { "vendor_id": "15B3", "device_id": "1013",
```
It's a bit scary that the NIC data from ethtool has a ```version``` and a ```firmware-version```. Hopefully you don't get NICs from the same manufacturer, with the same device IDs, but different ```versions``` (hardware I guess)  and different firmware-versions corresponding to the ```version```. I guess cross that bridge if it comes to it!

jovial replied:
Good point, should we add a version field to the matching rule?

dougszumski replied:
Hmm. +1 for leaving it as it is, and checking it later if we discover it's an issue.

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+
+        if _is_nic_verification_disabled(node):
+            LOG.warning('NIC firmware version verification has been disabled.')
+            return False
```
Should this return True so that cleaning passes when the verification is disabled?

jovial replied:
I believe returning `False` will still pass:

```
Any hardware manager that returns a result from the method call is considered a success and its return value passed on to whatever dispatched the method
```

but I am returning a list of successful checks later on (so should do that here)

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+            interface_name = interface.name
+            pci_id = interface.pci_id
+            if pci_id not in pci_matcher_lookup_table:
+                LOG.warning(_MISSING_MATCHER_RULE
```
In some ways I still wonder if we should fail here, because if you made a mistake with the ```device_id```  it would look like NIC firmware version checking was working when it actually wasn't. Since you can't easily see the warning you could imagine people being oblivious to it.

Perhaps the SE linux style mode flag that you suggested would be useful here so that the plugin could be configured to fail cleaning on warnings. 

I suppose if the additional cleaning plugin for verifying device trees were present we could avoid this issue, if it used the config from this plugin.

@markgoddard, @oneswig  - any more thoughts?

jovial replied:
I believe the device tree checker should run with a higher priority. If you ensured that the same the same ids were being used (shared config) then this would detect that case before it is an issue?

dougszumski replied:
Yeah - i think so long as we *ensure* the same ids are being used there's nothing to worry about.

markgoddard replied:
I think this is ok, as long as we catch the case (and fail) where a specified matcher does not match any NICs.

jovial replied:
This is now implemented.

dougszumski reviewed (changes requested):
Looks good to me overall. Some minor nits / questions inline.

dougszumski commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+            ]
+
+        The nic_firmware property can be manually set on a given node, as shown
+        in this example::
```
Please could we update the README in the root of the git repo so that it contains some instructions on how to use this.

jovial committed:
add system_nic moudle to README

jovial committed:
address some issues from code review

dougszumski reviewed (approved):
LGTM.

markgoddard commented on README.rst:
```diff
@@ -60,6 +60,49 @@ or:
+      {
+        "vendor_id": "<vendor_id>",
+        "device_id": "<device_id>",
+        "firmware_version": "<firmware_version>"
```
If I want to use your plugin, how would I find out this information about a NIC?

dougszumski replied:
Should we make sure that this data is collected by inspection?

markgoddard replied:
I still don't think that we should be automatically applying information found via inspection to the verification rules - it could open up a security hole.

dougszumski replied:
Agreed that it shouldn't be automatic. I was just thinking that a user could manually look at the inspection data to get an idea of what the dict should look like. They'd need to manually validate it / set which BIOS version they actually wanted.

jovial committed:
add some known limitations to readme

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+        split = itertools.takewhile(lambda x: x != "=", line)
+        key = "".join(split)
+        value = "".join(line)
+        result[interface][key] = value
```
Won't we end up with the key in the value here? e.g. "PCI_SLOT_NAME=0000:00:19.0"

jovial replied:
No, since the iterator is partially exhausted by `takewhile`, it just contains the remainder. Maybe line wasn't the best name...

markgoddard replied:
Ok, that makes sense.

markgoddard commented on stackhpc_ipa_hardware_managers/system_nic.py:
```diff
@@ -0,0 +1,367 @@
+        interfaces = self.get_interface_descriptors()
+        successes = {}
+        failures = {}
+        assert type(self.get_interface_descriptors()) == list
```
Is this left over debug?

jovial replied:
Would you remove the assert, or stop it duplicating the call and leave it in?

markgoddard replied:
I don't mind too much, it just stood out as the only assert in the code, checking something quite fundamental to a method in the same class. I guess the duplicate method call was a bit odd too.

markgoddard reviewed (commented):
This is looking good.

jovial committed:
clarify how to determine pci id and firmware version for system_nic

jovial committed:
address some code review issues

jovial committed:
tweak readability

jovial committed:
Clean now fails if any of the rules fail to match

markgoddard reviewed (commented):
We discussed summarising the 'rules' of the plugin in standup. I'd still quite like to see those, as a way of defining what the code does, and making it possible to verify that:

* we agree on what it should be doing
* it is doing what it should be doing

It's easy to get bogged down in the parsing of /proc files and ignore the core logic that will make this work (or not).

dougszumski commented:
When this is merged we should update all the nodes in the alt-1 and prod clusters to use it.

jovial commented:
ping - this has been open for some time. Is it something we still want?
